units: copy all needed `conftest.py`'s from parent dirs into collections

The problem
===========

`migrate.py` doesn't copy the `conftest` modules from upper-level directories as it should.

Complete example (one of them):
```python-traceback
______ ERROR at setup of test_xenserver_guest_powerstate_wait[wait-other] ______
[gw2] linux -- Python 3.7.4 /tmp/python-rndnyy_e-ansible/python
file /root/ansible/ansible_collections/ansible/misc/tests/unit/modules/cloud/xenserver/test_xenserver_guest_powerstate.py, line 250
  @pytest.mark.parametrize('patch_ansible_module',
                           testcase_module_params_wait['params'],
                           ids=testcase_module_params_wait['ids'],
                           indirect=True)
  def test_xenserver_guest_powerstate_wait(mocker, patch_ansible_module, capfd, XenAPI, xenserver_guest_powerstate):
E       fixture 'patch_ansible_module' not found
>       available fixtures: XenAPI, bigip, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, doctest_namespace, fake_ansible_module, mgmt_root, mock, mocker, monkeypatch, opt_bigip, opt_password, opt_peer, opt_port, opt_release, opt_token, opt_username, opt_vcmp_host, peer, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory, vcmp_host, worker_id, xenserver_guest_info, xenserver_guest_powerstate
>       use 'pytest --fixtures [testpath]' for help on them.

/root/ansible/ansible_collections/ansible/misc/tests/unit/modules/cloud/xenserver/test_xenserver_guest_powerstate.py:250
```


Easy solution
-------------

Walk the fs dir layout up the tree and copy all `conftest.py` modules from parent dirs.


Advanced solution
-----------------

Analyze FST of test modules and record all the fixture names. Then, scan `conftest.py` files and copy ones declaring the corresponding fixtures.


Bonus complexity
----------------

We could also analyze imports in `conftest.py` files in order to identify helper utils modules imports. But that's a problem described in #80.

--- a/migrate.py
+++ b/migrate.py
@@ -696,12 +696,58 @@ def copy_unit_tests(checkout_path, collection_dir, plugin_type, plugin, spec):
 
     # Find all test modules with the same ending as the current plugin
     plugin_dir, plugin_mod = os.path.split(plugin)
-    matching_test_modules = glob.glob(os.path.join(type_base_subdir, plugin_dir, f'*{plugin_mod}'))
+    matching_test_modules = set(glob.glob(os.path.join(
+        type_base_subdir,
+        plugin_dir,
+        f'*{plugin_mod}',
+    )))
     # Path(matching_test_modules[0]).relative_to(Path(checkout_path))
     # os.path.relpath(matching_test_modules[0], checkout_path)
     if not matching_test_modules:
         logger.info('No tests matching %s/%s found', plugin_type, plugin)
         return copy_map
+
+    def find_up_the_tree(target_path):
+        """Find conftest.py in the parent test dirs."""
+        needle_filename = 'conftest.py'
+        tests_root = os.path.join(checkout_path, unit_tests_relative_root)
+
+        relative_target_path = os.path.relpath(target_path, tests_root)
+        if relative_target_path.startswith('..'):
+            raise ValueError(
+                f'`{target_path}` is not a part of `{tests_root}`',
+            )
+
+        logger.info(
+            'Locating parent %s '
+            'for %s...',
+            needle_filename,
+            target_path,
+        )
+        while relative_target_path:
+            relative_target_path, _ = os.path.split(relative_target_path)
+
+            target_file = os.path.join(
+                tests_root,
+                relative_target_path,
+                needle_filename,
+            )
+            if not os.path.isfile(target_file):
+                continue
+
+            logger.info(
+                'Located %s...',
+                target_file
+            )
+            yield target_file
+
+    # Augment constest.py's from
+    # parent dirs:
+    matching_test_modules.update(
+        p
+        for m in matching_test_modules.copy()
+        for p in find_up_the_tree(m)
+    )
 
     def traverse_dir(path, relative_to):
         rel_path = os.path.join(relative_to, path)

